File: app/src/main/java/com/kartsev/dmitry/githubseeker/view/ui/adapters/recycler/BaseAdapterPagination.java
package com.kartsev.dmitry.githubseeker.view.ui.adapters.recycler;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;

import com.kartsev.dmitry.githubseeker.utils.LogUtils;

import java.util.List;

public abstract class BaseAdapterPagination<T> extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    public static final String LOG_TAG = "BaseAdapterPagination";
    protected List<T> list;
    private T loadingFooter;

    /*
        Pagination values
     */
    protected static final int ITEM = 0;
    protected static final int LOADING = 1;
    private boolean isLoadingAdded = false;

    public BaseAdapterPagination(List<T> list) {
        this.list = list;
    }

    public List<T> getList() {
        return list;
    }

    public void setList(List<T> list) {
        this.list = list;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public abstract RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType);

    @Override
    public int getItemCount() {
        return list.size();
    }

    private T getItem(int position) {
        return list.get(position);
    }

    @Override
    public int getItemViewType(int position) {
        return (position == list.size() - 1 && isLoadingAdded) ? LOADING : ITEM;
    }

    /*
       Helpers
    */

    public void setLoadingFooter(T loadingFooter) {
        this.loadingFooter = loadingFooter;
    }

    public void add(T item) {
        list.add(item);
    }

    public void addAll(List<T> itemsList) {
        LogUtils.LOGD(LOG_TAG, "adding " + itemsList);
        for (T item: itemsList) {
            add(item);
        }
        notifyDataSetChanged();
    }

    public void remove(T item) {
        int position = list.indexOf(item);
        if (position >= 0) {
            list.remove(position);
            notifyItemRemoved(position);
        }
    }

    public void clear() {
        isLoadingAdded = false;
        while (getItemCount() > 0) {
            remove(getItem(0));
        }
    }

    public boolean isEmpty() {
        return getItemCount() == 0;
    }

    public void addLoadingFooter() {
        isLoadingAdded = true;
        list.add(loadingFooter);
    }

    public void removeLoadingFooter() {
        isLoadingAdded = false;

        if (list.contains(loadingFooter)) {
            int position = list.indexOf(loadingFooter);
            list.remove(position);
            notifyItemRemoved(position);
        }
    }

    /**
     * Main list's content ViewHolder
     */
    protected abstract class ItemVH extends RecyclerView.ViewHolder {

        public ItemVH(View itemView) {
            super(itemView);
        }
    }

    /**
     * ILoading ViewHolder
     */
    protected abstract class LoadingVH extends RecyclerView.ViewHolder {

        public LoadingVH(View itemView) {
            super(itemView);
        }
    }
}

File: app/src/main/java/com/kartsev/dmitry/githubseeker/view/ui/adapters/recycler/BaseAdapterPagination.kt
package com.kartsev.dmitry.githubseeker.view.ui.adapters.recycler

import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import com.kartsev.dmitry.githubseeker.R

abstract class BaseAdapterPagination<T>(protected var mList: MutableList<T>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
    private var loadingFooter: T? = null
    private var isLoadingAdded = false

    val isEmpty: Boolean
        get() = itemCount == 0

    fun getList(): List<T> {
        return mList
    }

    fun setList(list: MutableList<T>) {
        this.mList = list
        notifyDataSetChanged()
    }

    abstract override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder

    override fun getItemCount(): Int {
        return mList.size
    }

    private fun getItem(position: Int): T {
        return mList[position]
    }

    override fun getItemViewType(position: Int): Int {
        return if (position == mList.size - 1 && isLoadingAdded) LOADING else ITEM
    }

    /*
       Helpers
    */

    fun setLoadingFooter(loadingFooter: T) {
        this.loadingFooter = loadingFooter
    }

    fun add(item: T) {
        mList.add(item)
    }

    fun addAll(itemsList: List<T>) {
        for (item in itemsList) {
            add(item)
        }
        notifyDataSetChanged()
    }

    fun remove(item: T) {
        val position = mList.indexOf(item)
        if (position >= 0) {
            mList.removeAt(position)
            notifyItemRemoved(position)
        }
    }

    fun clear() {
        isLoadingAdded = false
        while (itemCount > 0) {
            remove(getItem(0))
        }
    }

    fun addLoadingFooter() {
        isLoadingAdded = true
        loadingFooter?.let { mList.add(it) }
    }

    fun removeLoadingFooter() {
        isLoadingAdded = false

        if (mList.contains(loadingFooter)) {
            val position = mList.indexOf(loadingFooter)
            mList.removeAt(position)
            notifyItemRemoved(position)
        }
    }

    /**
     * Main list's content ViewHolder
     */
    abstract inner class ItemVH(itemView: View) : RecyclerView.ViewHolder(itemView)

    /**
     * ILoading ViewHolder
     */
    abstract inner class LoadingVH(itemView: View) : RecyclerView.ViewHolder(itemView)

    companion object {

        /*
        Pagination values
     */
        val ITEM = 0
        val LOADING = 1
    }
}
